File: backend/OcppBackend/Models/OcppModels.cs
// Models/OcppModels.cs
namespace OcppBackend.Models;

// ─── Enums ────────────────────────────────────────────────────────────────────

public enum ConnectorStatus
{
    Available,
    Preparing,
    Charging,
    SuspendedEVSE,
    SuspendedEV,
    Finishing,
    Reserved,
    Unavailable,
    Faulted
}

public enum ChargePointErrorCode
{
    NoError,
    ConnectorLockFailure,
    EVCommunicationError,
    GroundFailure,
    HighTemperature,
    InternalError,
    LocalListConflict,
    OtherError,
    OverCurrentFailure,
    PowerMeterFailure,
    PowerSwitchFailure,
    ReaderFailure,
    ResetFailure,
    UnderVoltage,
    OverVoltage,
    WeakSignal
}

public enum AuthorizationStatus
{
    Accepted,
    Blocked,
    Expired,
    Invalid,
    ConcurrentTx
}

public enum TransactionStatus
{
    Active,
    Completed,
    Invalid
}

public enum RegistrationStatus
{
    Accepted,
    Pending,
    Rejected
}

// ─── Database Entities ────────────────────────────────────────────────────────

public class ChargePoint
{
    public int Id { get; set; }
    public string ChargePointId { get; set; } = "";      // e.g. "CP-001"
    public string Vendor { get; set; } = "";
    public string Model { get; set; } = "";
    public string SerialNumber { get; set; } = "";
    public string FirmwareVersion { get; set; } = "";
    public bool IsOnline { get; set; }
    public DateTime? LastHeartbeat { get; set; }
    public DateTime? LastBootTime { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public ICollection<Connector> Connectors { get; set; } = new List<Connector>();
    public ICollection<Transaction> Transactions { get; set; } = new List<Transaction>();
}

public class Connector
{
    public int Id { get; set; }
    public int ChargePointId { get; set; }
    public int ConnectorId { get; set; }               // 1-based per OCPP spec
    public ConnectorStatus Status { get; set; } = ConnectorStatus.Unavailable;
    public ChargePointErrorCode ErrorCode { get; set; } = ChargePointErrorCode.NoError;
    public string? ErrorInfo { get; set; }
    public DateTime? StatusTimestamp { get; set; }

    [System.Text.Json.Serialization.JsonIgnore]
    public ChargePoint ChargePoint { get; set; } = null!;
    public ICollection<Transaction> Transactions { get; set; } = new List<Transaction>();
}

public class IdTag
{
    public int Id { get; set; }
    public string TagId { get; set; } = "";            // RFID tag string
    public string? UserId { get; set; }                // linked user (optional)
    public string? UserName { get; set; }
    public AuthorizationStatus Status { get; set; } = AuthorizationStatus.Accepted;
    public DateTime? ExpiryDate { get; set; }
    public string? ParentTagId { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public string? Note { get; set; }
}

public class Transaction
{
    public int Id { get; set; }
    public int TransactionId { get; set; }             // OCPP transaction ID
    public int ChargePointId { get; set; }
    public int ConnectorId { get; set; }               // FK to Connector.Id
    public int ConnectorNumber { get; set; }           // 1-based connector number
    public string IdTag { get; set; } = "";
    public TransactionStatus Status { get; set; } = TransactionStatus.Active;
    public int MeterStart { get; set; }                // Wh
    public int? MeterStop { get; set; }                // Wh
    public DateTime StartTime { get; set; }
    public DateTime? StopTime { get; set; }
    public string? StopReason { get; set; }
    public double? EnergyDeliveredKwh { get; set; }

    [System.Text.Json.Serialization.JsonIgnore]
    public ChargePoint ChargePoint { get; set; } = null!;

    [System.Text.Json.Serialization.JsonIgnore]
    public Connector Connector { get; set; } = null!;
    public ICollection<MeterValue> MeterValues { get; set; } = new List<MeterValue>();
}

public class MeterValue
{
    public int Id { get; set; }
    public int TransactionId { get; set; }             // FK to Transaction.Id
    public DateTime Timestamp { get; set; }
    public double? EnergyWh { get; set; }
    public double? PowerW { get; set; }
    public double? VoltageV { get; set; }
    public double? CurrentA { get; set; }
    public string? Context { get; set; }               // "Sample.Periodic" etc.

    [System.Text.Json.Serialization.JsonIgnore]
    public Transaction Transaction { get; set; } = null!;
}

public class OcppMessageLog
{
    public int Id { get; set; }
    public string ChargePointId { get; set; } = "";
    public string Direction { get; set; } = "";        // "CP->CSMS" or "CSMS->CP"
    public string Action { get; set; } = "";
    public string MessageId { get; set; } = "";
    public string Payload { get; set; } = "";          // raw JSON
    public bool IsError { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
}

// ─── OCPP 1.6 Message Structures (used inside hub/service) ────────────────────

public record OcppBootNotificationRequest(
    string ChargePointVendor,
    string ChargePointModel,
    string? ChargePointSerialNumber,
    string? ChargeBoxSerialNumber,
    string? FirmwareVersion,
    string? Iccid,
    string? Imsi
);

public record OcppBootNotificationResponse(
    string Status,           // "Accepted" | "Pending" | "Rejected"
    string CurrentTime,
    int Interval
);

public record OcppHeartbeatResponse(string CurrentTime);

public record OcppIdTagInfo(
    string Status,
    string? ExpiryDate,
    string? ParentIdTag
);

public record OcppAuthorizeRequest(string IdTag);
public record OcppAuthorizeResponse(OcppIdTagInfo IdTagInfo);

public record OcppStartTransactionRequest(
    int ConnectorId,
    string IdTag,
    int MeterStart,
    string Timestamp,
    int? ReservationId
);

public record OcppStartTransactionResponse(
    int TransactionId,
    OcppIdTagInfo IdTagInfo
);

public record OcppStopTransactionRequest(
    int TransactionId,
    string? IdTag,
    int MeterStop,
    string Timestamp,
    string? Reason,
    List<object>? TransactionData
);

public record OcppStopTransactionResponse(OcppIdTagInfo? IdTagInfo);

public record OcppStatusNotificationRequest(
    int ConnectorId,
    string ErrorCode,
    string Status,
    string? Info,
    string? Timestamp,
    string? VendorId,
    string? VendorErrorCode
);

public record OcppMeterValuesRequest(
    int ConnectorId,
    int? TransactionId,
    List<OcppMeterValueEntry> MeterValue
);

public record OcppMeterValueEntry(
    string Timestamp,
    List<OcppSampledValue> SampledValue
);

public record OcppSampledValue(
    string Value,
    string? Context,
    string? Format,
    string? Measurand,
    string? Phase,
    string? Location,
    string? Unit
);

// ─── Remote Commands (CSMS → CP) ─────────────────────────────────────────────

public record RemoteStartTransactionRequest(string IdTag, int? ConnectorId, object? ChargingProfile);
public record RemoteStopTransactionRequest(int TransactionId);
public record ChangeAvailabilityRequest(int ConnectorId, string Type);  // "Operative"|"Inoperative"
public record ResetRequest(string Type);                                  // "Hard"|"Soft"
public record UnlockConnectorRequest(int ConnectorId);
public record GetConfigurationRequest(List<string>? Key);
public record ChangeConfigurationRequest(string Key, string Value);
public record TriggerMessageRequest(string RequestedMessage, int? ConnectorId);
public record ClearCacheRequest();

// ─── CSMS → App notification payloads (via SignalR to Flutter apps) ───────────

public record ChargePointStatusUpdate(
    string ChargePointId,
    bool IsOnline,
    DateTime? LastHeartbeat,
    IEnumerable<ConnectorStatusDto> Connectors
);

public record ConnectorStatusDto(
    int ConnectorId,
    string Status,
    string ErrorCode,
    int? ActiveTransactionId
);

public record TransactionStartedEvent(
    int TransactionId,
    string ChargePointId,
    int ConnectorId,
    string IdTag,
    DateTime StartTime,
    int MeterStart
);

public record TransactionUpdatedEvent(
    int TransactionId,
    double EnergyWh,
    double? PowerW,
    double? VoltageV,
    double? CurrentA,
    DateTime Timestamp
);

public record TransactionStoppedEvent(
    int TransactionId,
    string ChargePointId,
    int ConnectorId,
    string IdTag,
    DateTime StopTime,
    int MeterStop,
    double EnergyDeliveredKwh,
    string? Reason
);

File: standalone/lib/models/ocpp_models.dart
// lib/models/ocpp_models.dart

enum OcppVersion { ocpp16, ocpp201 }

enum ChargePointStatus {
  available,
  preparing,
  charging,
  suspendedEVSE,
  suspendedEV,
  finishing,
  reserved,
  unavailable,
  faulted,
}

enum ConnectorStatus {
  available,
  occupied,
  reserved,
  unavailable,
  faulted,
}

enum AuthorizationStatus {
  accepted,
  blocked,
  expired,
  invalid,
  concurrentTx,
}

enum TransactionStatus { idle, authorized, charging, stopping }

enum OcppMessageType { call, callResult, callError }

enum ChargePointErrorCode {
  connectorLockFailure,
  evCommunicationError,
  groundFailure,
  highTemperature,
  internalError,
  localListConflict,
  noError,
  otherError,
  overCurrentFailure,
  powerMeterFailure,
  powerSwitchFailure,
  readerFailure,
  resetFailure,
  underVoltage,
  overVoltage,
  weakSignal,
}

// ─── OCPP Message Envelope ───────────────────────────────────────────────────

class OcppMessage {
  final OcppMessageType type;
  final String messageId;
  final String? action;       // only for CALL
  final dynamic payload;
  final String? errorCode;    // only for CALLERROR
  final String? errorDescription;

  const OcppMessage({
    required this.type,
    required this.messageId,
    this.action,
    this.payload,
    this.errorCode,
    this.errorDescription,
  });

  List<dynamic> toJson() {
    switch (type) {
      case OcppMessageType.call:
        return [2, messageId, action, payload ?? {}];
      case OcppMessageType.callResult:
        return [3, messageId, payload ?? {}];
      case OcppMessageType.callError:
        return [4, messageId, errorCode ?? 'GenericError', errorDescription ?? '', {}];
    }
  }
}

// ─── Connector Model ─────────────────────────────────────────────────────────

class ConnectorModel {
  final int id;
  ConnectorStatus status;
  double currentPowerKw;
  double energyDeliveredKwh;
  String? activeTransactionId;
  String? authorizedIdTag;
  DateTime? sessionStart;
  List<MeterSample> meterHistory;

  ConnectorModel({
    required this.id,
    this.status = ConnectorStatus.available,
    this.currentPowerKw = 0,
    this.energyDeliveredKwh = 0,
    this.activeTransactionId,
    this.authorizedIdTag,
    this.sessionStart,
    List<MeterSample>? meterHistory,
  }) : meterHistory = meterHistory ?? [];

  Duration get sessionDuration =>
      sessionStart != null ? DateTime.now().difference(sessionStart!) : Duration.zero;

  double get sessionCostEstimate => energyDeliveredKwh * 0.30; // $0.30/kWh mock

  ConnectorModel copyWith({
    ConnectorStatus? status,
    double? currentPowerKw,
    double? energyDeliveredKwh,
    String? activeTransactionId,
    String? authorizedIdTag,
    DateTime? sessionStart,
    List<MeterSample>? meterHistory,
  }) {
    return ConnectorModel(
      id: id,
      status: status ?? this.status,
      currentPowerKw: currentPowerKw ?? this.currentPowerKw,
      energyDeliveredKwh: energyDeliveredKwh ?? this.energyDeliveredKwh,
      activeTransactionId: activeTransactionId ?? this.activeTransactionId,
      authorizedIdTag: authorizedIdTag ?? this.authorizedIdTag,
      sessionStart: sessionStart ?? this.sessionStart,
      meterHistory: meterHistory ?? this.meterHistory,
    );
  }
}

// ─── Meter Value Sample ───────────────────────────────────────────────────────

class MeterSample {
  final DateTime timestamp;
  final double powerKw;
  final double energyKwh;
  final double voltage;
  final double currentA;

  const MeterSample({
    required this.timestamp,
    required this.powerKw,
    required this.energyKwh,
    required this.voltage,
    required this.currentA,
  });
}

// ─── Charge Point Configuration ──────────────────────────────────────────────

class ChargePointConfig {
  String chargePointId;
  String centralSystemUrl;
  OcppVersion ocppVersion;
  int numberOfConnectors;
  String vendor;
  String model;
  String serialNumber;
  String firmwareVersion;
  int heartbeatInterval;       // seconds
  int meterValueInterval;      // seconds
  double maxPowerKw;

  ChargePointConfig({
    this.chargePointId = 'CP-SIMULATOR-001',
    this.centralSystemUrl = 'ws://localhost:9000/CP-SIMULATOR-001',
    this.ocppVersion = OcppVersion.ocpp16,
    this.numberOfConnectors = 2,
    this.vendor = 'SimCo',
    this.model = 'SimCharger X1',
    this.serialNumber = 'SIM-2024-00001',
    this.firmwareVersion = '1.0.0',
    this.heartbeatInterval = 30,
    this.meterValueInterval = 15,
    this.maxPowerKw = 22.0,
  });
}

// ─── Log Entry ────────────────────────────────────────────────────────────────

enum LogDirection { sent, received, system }

class LogEntry {
  final DateTime timestamp;
  final LogDirection direction;
  final String message;
  final String? rawData;
  final bool isError;

  const LogEntry({
    required this.timestamp,
    required this.direction,
    required this.message,
    this.rawData,
    this.isError = false,
  });
}

// ─── Connection State ─────────────────────────────────────────────────────────

enum CpConnectionState { disconnected, connecting, connected, reconnecting }